patch_subset/cbor/client_state.cc

```diff
@@ -0,0 +1,147 @@
+#include "patch_subset/cbor/client_state.h"
+
+#include "common/status.h"
+#include "patch_subset/cbor/cbor_utils.h"
+#include "patch_subset/cbor/compressed_int_list.h"
+
+namespace patch_subset::cbor {
+
+using std::string;
+using std::vector;
+
+ClientState::ClientState()
+    : _font_id(std::nullopt),
+      _font_data(std::nullopt),
+      _fingerprint(std::nullopt),
+      _codepoint_remapping(std::nullopt) {}
+
+ClientState::ClientState(const string& font_id, const string& font_data,
+                         uint64_t fingerprint,
+                         const vector<int32_t>& codepoint_remapping)
+    : _font_id(font_id),
+      _font_data(font_data),
+      _fingerprint(fingerprint),
+      _codepoint_remapping(codepoint_remapping) {}
+
+StatusCode ClientState::Decode(const cbor_item_t& cbor_map, ClientState& out) {
+  ClientState result;
+  if (!cbor_isa_map(&cbor_map) || cbor_map_is_indefinite(&cbor_map)) {
+    return StatusCode::kInvalidArgument;
+  }
+  StatusCode sc =
+      CborUtils::GetStringField(cbor_map, kFontIdFieldNumber, result._font_id);
+  if (sc != StatusCode::kOk) {
+    return StatusCode::kInvalidArgument;
+  }
+  sc = CborUtils::GetBytesField(cbor_map, kFontDataFieldNumber,
+                                result._font_data);
+  if (sc != StatusCode::kOk) {
+    return StatusCode::kInvalidArgument;
+  }
+  sc = CborUtils::GetUInt64Field(cbor_map, kFingerprintFieldNumber,
+                                 result._fingerprint);
+  if (sc != StatusCode::kOk) {
+    return StatusCode::kInvalidArgument;
+  }
+  sc = CompressedIntList::GetIntListField(
+      cbor_map, kCodepointRemappingFieldNumber, result._codepoint_remapping);
+  if (sc != StatusCode::kOk) {
+    return StatusCode::kInvalidArgument;
+  }
+  out = result;
```

Comment:
This will result in copying the fontdata field, which could be sizable. Can you add a swap() method to ClientState which internally will use swap() on the fontdata string. Or alternatively implement a move constructor for ClientState and use std::move() here.

Reply:
I added a TODO, I think this is beyond the scope of this PR.